src/ring_theory/power_basis.lean

```diff
@@ -111,6 +117,214 @@ begin
+begin
+  rw [← aeval_mod_by_monic_eq_self_of_root
+          (minimal_polynomial.monic pb.is_integral_gen)
+          (minimal_polynomial.aeval _),
+      ← @aeval_mod_by_monic_eq_self_of_root _ _ _ _ _ f _
+          (minimal_polynomial.monic pb.is_integral_gen) y hy],
+  by_cases hf : f %ₘ minimal_polynomial (pb.is_integral_gen) = 0,
+  { simp only [hf, alg_hom.map_zero, linear_map.map_zero] },
+  have : (f %ₘ minimal_polynomial _).nat_degree < pb.dim,
+  { rw ← pb.nat_degree_minimal_polynomial,
+    apply nat_degree_lt_nat_degree hf,
+    exact degree_mod_by_monic_lt _ (minimal_polynomial.monic _) (minimal_polynomial.ne_zero _) },
+  rw [aeval_eq_sum_range' this, aeval_eq_sum_range' this, linear_map.map_sum],
+  refine finset.sum_congr rfl (λ i (hi : i ∈ finset.range pb.dim), _),
+  rw finset.mem_range at hi,
+  rw linear_map.map_smul,
+  congr,
+  exact @constr_basis _ _ _ _ _ _ _ _ _ _ _ (⟨i, hi⟩ : fin pb.dim) pb.is_basis,
+end
+
+lemma constr_pow_gen (pb : power_basis A S) {y : S'}
+  (hy : aeval y (minimal_polynomial pb.is_integral_gen) = 0) :
+  pb.is_basis.constr (λ i, y ^ (i : ℕ)) pb.gen = y :=
+by { convert pb.constr_pow_aeval hy X; rw aeval_X }
+
+lemma constr_pow_algebra_map (pb : power_basis A S) {y : S'}
+  (hy : aeval y (minimal_polynomial pb.is_integral_gen) = 0) (x : A) :
+  pb.is_basis.constr (λ i, y ^ (i : ℕ)) (algebra_map A S x) = algebra_map A S' x :=
+by { convert pb.constr_pow_aeval hy (C x); rw aeval_C }
+
+lemma constr_pow_mul [nontrivial S] (pb : power_basis A S) {y : S'}
+  (hy : aeval y (minimal_polynomial pb.is_integral_gen) = 0) (x x' : S) :
+  pb.is_basis.constr (λ i, y ^ (i : ℕ)) (x * x') =
+    pb.is_basis.constr (λ i, y ^ (i : ℕ)) x * pb.is_basis.constr (λ i, y ^ (i : ℕ)) x' :=
+begin
+  obtain ⟨f, hf, rfl⟩ := pb.exists_eq_aeval x,
+  obtain ⟨g, hg, rfl⟩ := pb.exists_eq_aeval x',
+  simp only [← aeval_mul, pb.constr_pow_aeval hy]
+end
+
+/-- `pb.lift y hy` is the algebra map sending `pb.gen` to `y`,
+where `hy` states the higher powers of `y` are the same as the higher powers of `pb.gen`. -/
+noncomputable def lift [nontrivial S] (pb : power_basis A S) (y : S')
+  (hy : aeval y (minimal_polynomial pb.is_integral_gen) = 0) :
+  S →ₐ[A] S' :=
+{ map_one' := by { convert pb.constr_pow_algebra_map hy 1 using 2; rw ring_hom.map_one },
+  map_zero' := by { convert pb.constr_pow_algebra_map hy 0 using 2; rw ring_hom.map_zero },
+  map_mul' := pb.constr_pow_mul hy,
+  commutes' := pb.constr_pow_algebra_map hy,
+  .. pb.is_basis.constr (λ i, y ^ (i : ℕ)) }
+
+@[simp] lemma lift_gen [nontrivial S] (pb : power_basis A S) (y : S')
+  (hy : aeval y (minimal_polynomial pb.is_integral_gen) = 0) :
+  pb.lift y hy pb.gen = y :=
+pb.constr_pow_gen hy
+
+@[simp] lemma lift_aeval [nontrivial S] (pb : power_basis A S) (y : S')
+  (hy : aeval y (minimal_polynomial pb.is_integral_gen) = 0) (f : polynomial A) :
+  pb.lift y hy (aeval pb.gen f) = aeval y f :=
+pb.constr_pow_aeval hy f
```

Comment:
Is there an `equiv` hiding here of the form `S' ≃ (S →ₐ[A] S')`, where `inv_fun := λ F, f pb.gen`?

Reply:
I don't think so, since not all `y : S'` satisfy `(hy : aeval y (minimal_polynomial pb.is_integral_gen) = 0)`. (And for finite-dimensional field extensions, we have that there are only finitely many `alg_hom`s but the fields can have infinitely many elements.) Or do I misinterpret the suggestion?

Reply:
Maybe `{y : S' // aeval y (minimal_polynomial pb.is_integral_gen) = 0) ≃ (S →ₐ[A] S')`?

I don't really know the background here - I've just noticed that whenever someone uses `lift` in their function name, they've defined an equivalence of some kind.

Reply:
Mapping them to the roots is indeed an equiv, and in fact PR #4831 [contains the case where S is a simple field extension of A.](https://github.com/leanprover-community/mathlib/pull/4831/files#diff-a4131bb7c0e21cfcee5d0ec2868cf304a9ce8acad1e543265886315364d4a75eR419)

Reply:
What's your plan here WRT that PR? Should this one wait for that one, or vice versa? Do you want to update yours to use the equiv?

Reply:
In principle, the two PRs are independent and neither is exactly an extension of the other. So I would like to propose that we merge them separately and create a new PR for turning `lift` into an equiv on all power bases by adapting the code in #4831.